File: test/test_security_audit.mjs
/**
 * Avinya Care Foundation - Automated Security & RBAC Audit Test Suite
 */

import http from 'http';

const BASE_URL = 'http://localhost:3000';

function makeRequest(path, options = {}, body = null) {
  return new Promise((resolve, reject) => {
    const url = new URL(path, BASE_URL);
    const reqOptions = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname + url.search,
      method: options.method || 'GET',
      headers: options.headers || {}
    };

    const req = http.request(reqOptions, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        let json = null;
        try { json = JSON.parse(data); } catch (e) {}
        resolve({ status: res.statusCode, headers: res.headers, body: json, raw: data });
      });
    });

    req.on('error', reject);
    if (body) {
      req.write(typeof body === 'string' ? body : JSON.stringify(body));
    }
    req.end();
  });
}

async function runSecurityAuditTests() {
  console.log('====================================================');
  console.log('  AVINYA CARE SECURITY & PRIVACY AUDIT TEST SUITE   ');
  console.log('====================================================\n');

  let passed = 0;
  let failed = 0;

  // Test 1: Diagnostic Booking Unauthenticated GET Leak Protection
  try {
    const res = await makeRequest('/api/diagnostic-booking.php');
    if (res.status === 401 || res.status === 403) {
      console.log(`✓ TEST 1 PASSED: Public GET /api/diagnostic-booking.php correctly rejected with HTTP ${res.status}.`);
      passed++;
    } else {
      console.error(`✗ TEST 1 FAILED: Expected 401 or 403, got ${res.status}`);
      failed++;
    }
  } catch (e) {
    console.error('✗ TEST 1 ERROR:', e.message);
    failed++;
  }

  // Test 2: Doctor Appointment Unauthenticated GET List Leak Protection
  try {
    const res = await makeRequest('/api/booking/index.php?action=list');
    if (res.status === 401 || res.status === 403) {
      console.log(`✓ TEST 2 PASSED: Public GET /api/booking/index.php?action=list correctly rejected with HTTP ${res.status}.`);
      passed++;
    } else {
      console.error(`✗ TEST 2 FAILED: Expected 401 or 403, got ${res.status}`);
      failed++;
    }
  } catch (e) {
    console.error('✗ TEST 2 ERROR:', e.message);
    failed++;
  }

  // Test 3: Account Enumeration Prevention on Forgot Password
  try {
    const res = await makeRequest('/api/admin-auth.php?action=forgot_password', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    }, { email: 'nonexistent.user.98765@gmail.com' });

    if (res.status === 200 && res.body?.message?.includes('If an account exists')) {
      console.log('✓ TEST 3 PASSED: Account enumeration prevented (generic response returned).');
      passed++;
    } else {
      console.error(`✗ TEST 3 FAILED: Expected generic 200 response, got status ${res.status}`, res.body);
      failed++;
    }
  } catch (e) {
    console.error('✗ TEST 3 ERROR:', e.message);
    failed++;
  }

  // Test 4: Unauthenticated Admin Action Escalation Rejection
  try {
    const res = await makeRequest('/api/admin-data.php', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' }
    }, { action: 'delete_user', id: 'usr-admin-01' });

    if (res.status === 401) {
      console.log('✓ TEST 4 PASSED: Unauthenticated admin manipulation rejected with 401 Unauthorized.');
      passed++;
    } else {
      console.error(`✗ TEST 4 FAILED: Expected 401, got ${res.status}`);
      failed++;
    }
  } catch (e) {
    console.error('✗ TEST 4 ERROR:', e.message);
    failed++;
  }

  // Test 5: Security Health Headers Check
  try {
    const res = await makeRequest('/api/security-health.php');
    const headers = res.headers;
    const hasNosniff = headers['x-content-type-options'] === 'nosniff';
    const hasFrameOptions = headers['x-frame-options'] === 'SAMEORIGIN';
    const hasPolicy = !!headers['permissions-policy'];

    if (res.status === 200 && hasNosniff && hasFrameOptions && hasPolicy) {
      console.log('✓ TEST 5 PASSED: Security monitoring endpoint active with full security headers.');
      passed++;
    } else {
      console.error('✗ TEST 5 FAILED: Security headers missing or incomplete.', res.headers);
      failed++;
    }
  } catch (e) {
    console.error('✗ TEST 5 ERROR:', e.message);
    failed++;
  }

  console.log('\n====================================================');
  console.log(`  RESULTS: ${passed} PASSED, ${failed} FAILED`);
  console.log('====================================================');

  if (failed > 0) process.exit(1);
}

runSecurityAuditTests();

File: test/test_security_audit.py
#!/usr/bin/env python3
"""
Avinya Care Foundation - Security & RBAC Audit Test Suite (Python)
"""

import sys
import json
import urllib.request
import urllib.error

BASE_URL = "http://localhost:3000"

def make_request(path, method="GET", headers=None, data=None):
    url = BASE_URL + path
    req_headers = headers or {}
    encoded_data = None
    if data:
        if isinstance(data, dict):
            encoded_data = json.dumps(data).encode('utf-8')
            if 'Content-Type' not in req_headers:
                req_headers['Content-Type'] = 'application/json'
        elif isinstance(data, str):
            encoded_data = data.encode('utf-8')

    req = urllib.request.Request(url, data=encoded_data, headers=req_headers, method=method)
    try:
        with urllib.request.urlopen(req) as response:
            resp_body = response.read().decode('utf-8')
            json_body = None
            try:
                json_body = json.loads(resp_body)
            except Exception:
                pass
            return response.status, dict(response.headers), json_body, resp_body
    except urllib.error.HTTPError as e:
        resp_body = e.read().decode('utf-8')
        json_body = None
        try:
            json_body = json.loads(resp_body)
        except Exception:
            pass
        return e.code, dict(e.headers), json_body, resp_body
    except Exception as e:
        return 0, {}, None, str(e)

def run_tests():
    print("====================================================")
    print("  AVINYA CARE SECURITY & PRIVACY AUDIT TEST SUITE   ")
    print("====================================================\n")

    passed = 0
    failed = 0

    # Test 1: Diagnostic Booking Unauthenticated GET Leak Protection
    status, headers, body, raw = make_request("/api/diagnostic-booking.php")
    if status in (401, 403):
        print(f"✓ TEST 1 PASSED: Public GET /api/diagnostic-booking.php correctly rejected with HTTP {status}.")
        passed += 1
    else:
        print(f"✗ TEST 1 FAILED: Expected 401 or 403, got {status}")
        failed += 1

    # Test 2: Doctor Appointment Unauthenticated GET List Leak Protection
    status, headers, body, raw = make_request("/api/booking/index.php?action=list")
    if status in (401, 403):
        print(f"✓ TEST 2 PASSED: Public GET /api/booking/index.php?action=list correctly rejected with HTTP {status}.")
        passed += 1
    else:
        print(f"✗ TEST 2 FAILED: Expected 401 or 403, got {status}")
        failed += 1

    # Test 3: Account Enumeration Prevention on Forgot Password
    status, headers, body, raw = make_request("/api/admin-auth.php?action=forgot_password", method="POST", data={"email": "nonexistent.user.98765@gmail.com"})
    if status == 200 and body and "If an account exists" in body.get("message", ""):
        print("✓ TEST 3 PASSED: Account enumeration prevented (generic response returned).")
        passed += 1
    else:
        print(f"✗ TEST 3 FAILED: Expected generic 200 response, got status {status}, body: {body}")
        failed += 1

    # Test 4: Unauthenticated Admin Action Escalation Rejection
    status, headers, body, raw = make_request("/api/admin-data.php", method="POST", data={"action": "delete_user", "id": "usr-admin-01"})
    if status == 401:
        print("✓ TEST 4 PASSED: Unauthenticated admin manipulation rejected with 401 Unauthorized.")
        passed += 1
    else:
        print(f"✗ TEST 4 FAILED: Expected 401, got {status}")
        failed += 1

    # Test 5: Security Health Headers Check
    status, headers, body, raw = make_request("/api/security-health.php")
    has_nosniff = headers.get("X-Content-Type-Options") == "nosniff" or headers.get("x-content-type-options") == "nosniff"
    has_frame = headers.get("X-Frame-Options") == "SAMEORIGIN" or headers.get("x-frame-options") == "SAMEORIGIN"
    if status == 200 and has_nosniff and has_frame:
        print("✓ TEST 5 PASSED: Security health endpoint active with required headers.")
        passed += 1
    else:
        print(f"✗ TEST 5 FAILED: Security health check failed. Status: {status}, Headers: {headers}")
        failed += 1

    print("\n====================================================")
    print(f"  RESULTS: {passed} PASSED, {failed} FAILED")
    print("====================================================")

    if failed > 0:
        sys.exit(1)

if __name__ == "__main__":
    run_tests()
